smqtk_iqr/iqr/iqr_session.py

```diff
@@ -300,25 +316,72 @@ def refine(self) -> None:
 
         """
         with self.lock:
-            # combine pos/neg adjudications + added external data descriptors
+            # Combine pos/neg adjudications + added external data descriptors
             pos = [desc.vector() for desc in (self.positive_descriptors |
                                               self.external_positive_descriptors)]
-
             neg = [desc.vector() for desc in (self.negative_descriptors |
                                               self.external_negative_descriptors)]
 
             if not pos:
                 raise RuntimeError("Did not find at least one positive "
                                    "adjudication.")
 
-            self._log.debug("Ranking working set with %d pos and %d neg total "
-                            "examples.", len(pos), len(neg))
+            # Get working set descriptors
             pool_uids, pool_de = zip(*self.working_set.items())
             pool = [de.vector() for de in pool_de]
+
+            # Auto-select negative examples if none are given
+            if not neg:
+                neg_autoselect = set()
+                self._log.info(f"Auto-selecting negative examples. \
+                              ({self.autoneg_select_ratio} per positive)")
+                # For each positive example, find the farthest descriptor
+                # from it to use as a negative example
+                for p in pos:
+                    # Compute distance between current positive example and
+                    # each of the working set descriptors
+                    distances = []
+                    for desc_v in pool:
+                        distances.append(self.distance_metric(p, np.array(desc_v)))
+
+                    np_distances: np.array = np.array(distances)
```

Comment:
It looks like the cosine_distance function is defined as `(1D-vec, 1D-vec) -> float`. I had written it with the for loop to handle this case, otherwise I get an error if I try to call it without the loop, like `np_distances = self.distance_metric(p, pool_mat)`.